xbmc/addons/AddonInfo.cpp

```diff
@@ -151,7 +414,448 @@ CAddonInfo::CAddonInfo(std::string id, TYPE type)
     m_mainType(type),
     m_packageSize(0)
 {
+  CAddonType addonType(m_mainType, this, nullptr);
+  m_types.push_back(addonType);
+}
+
+bool CAddonInfo::IsType(TYPE type) const
+{
+  return (m_mainType == type || ProvidesSubContent(type));
+}
+
+const CAddonType* CAddonInfo::Type(TYPE type) const
+{
+  if (type == ADDON_UNKNOWN)
+    return &m_types[0];
+
+  for (auto& addonType : m_types)
+  {
+    if (addonType.Type() == type)
+      return &addonType;
+  }
+  return nullptr;
+}
+
+std::string CAddonInfo::MainLibPath() const
+{
+  if (m_types.empty())
+    return "";
+  return m_types[0].LibPath();
+}
+
+bool CAddonInfo::ProvidesSubContent(const TYPE& content, const TYPE& mainType/* = ADDON_UNKNOWN*/) const
+{
+  if (content == ADDON_UNKNOWN)
+    return false;
+
+  for (auto addonType : m_types)
+  {
+    if ((mainType == ADDON_UNKNOWN || addonType.Type() == mainType) && addonType.ProvidesSubContent(content))
+      return true;
+  }
+
+  return false;
+}
+
+bool CAddonInfo::ProvidesSeveralSubContents() const
+{
+  int contents = 0;
+  for (auto addonType : m_types)
+    contents += addonType.ProvidedSubContents();
+  return contents > 0 ? true : false;
+}
+
+std::string CAddonInfo::SerializeMetadata()
```

Comment:
It is a from database used part complete right!

But I must say again, to have in Database is complete bad!

It use values from CAddonInfo and use only a string as "in" or "out" for database. All used values from the string are from CAddonInfo.

**In case now new values becomes added (removed bring God thanks to a compile error), is the database be easily forgotten!**

Further is CAddonInfo not a generic class, it is the base addon class (and on end only one time for one addon present) to declare the whole addon in Kodi.